jwst/dark_current/dark_sub.py

```diff
@@ -178,23 +275,38 @@ def subtract_dark(input, dark):
 
     """
 
+    dark_ints = dark.data.shape[0] -1
+
     log.debug("subtract_dark: nints=%d, ngroups=%d, size=%d,%d",
               input.meta.exposure.nints, input.meta.exposure.ngroups,
               input.data.shape[-1], input.data.shape[-2])
 
+
     # Create output as a copy of the input science data model
     output = input.copy()
 
+    
     # combine the science and dark DQ arrays
-    output.pixeldq = np.bitwise_or(input.pixeldq, dark.dq)
+    detector = input.meta.instrument.detector
+    if(detector[:3] == 'MIR'):
+           output.pixeldq = np.bitwise_or(input.pixeldq, np.bitwise_or(dark.dq[0,0,:,:], dark.dq[1,0,:,:]))
+    else:
+           output.pixeldq = np.bitwise_or(input.pixeldq, dark.dq)
 
 
     # loop over all integrations and groups in input science data
     for i in range(input.data.shape[0]):
-        for j in range(input.data.shape[1]):
+        if(detector[:3] =='MIR'):
+
+            if(i <= dark_ints): dark_int = dark.data[i]
```

Comment:
Howard,
Either way works. You are right  "<=" in line 301 is true for the first 
two integrations but that is fine because in the first integration
i = 0 so the dark used is dark_data[0] and in the second integration i = 
1 so the dark used is dark_data[1]

But I can see how that may not be straight forward thinking for others  
so I will make it cleaner (and easier to maintain)
and use the suggestions you sent.

For collapsing the dark dq planes - the np.bitwise_or can only take two 
arrays. I found a way to make it work, but I am still
trying to find a simplier numpy function. But as we mentioned earlier 
maybe the best way to handle this is have a single plane
dark DQ and let the IDT team figure out the best way to make this (me 
with my other hat on with help from other MIRI IDT members.

I am going to do a bit more testing and when I am happy I will commit 
the changes and make a pull request.
That is when you can see them - right ?

You are on vacation right ? You can just reply to this email when you 
get back - unless you would like this merged with JWST/master this week

Jane

On 07/25/2016 08:47 AM, Howard Bushouse wrote:
>
> In jwst/dark_current/dark_sub.py 
> <https://github.com/STScI-JWST/jwst/pull/118#discussion_r72089613>:
>
> >
> >
> >      # loop over all integrations and groups in input science data
> >      for i in range(input.data.shape[0]):
> > -        for j in range(input.data.shape[1]):
> > +        if(detector[:3] =='MIR'):
> > +
> > +            if(i <= dark_ints): dark_int = dark.data[i]
>
> This doesn't look quite right. Assuming the dark data contain 2 
> integrations, line 278 sets dark_ints to 1. On the first iteration of 
> the loop in line 298 i will be zero, and on the 2nd iteration it will 
> be 1. So the "<=" comparison in line 301 will be true for the first 
> TWO iterations of the loop, but you only want it to be true on the 
> first, so that unique dark data get used only for the first 
> integration of the science data. So I think line 301 should be:
>
> if(i < dark_ints):
>
> and line 302 should be:
>
> if(i >= dark_ints):
>
> Actually you don't even need that IF structure. It could be a simple 
> IF/ELSE, because the only unique case is the first integration:
>
> if (i < dark_ints):
> dark_int = dark.data[i]
> else:
> dark_int = dark.data[dark_ints]
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/STScI-JWST/jwst/pull/118/files/76105243e0379a5934adbb1034a1377a5d1cec05#r72089613>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/ARonWVkDeQBE_ZLSWIh-CRzkjsHPxNciks5qZNqogaJpZM4JSExt>.
>